adding fabricpath topo and fixed feature_set as a class method

dcheriancisco opened the pull request:
# FABRICPATH TOPLOGY:

Node under test:
- name  - n8k-136
- type  - N5K-C56128P
- image - bootflash:///n6000-uk9.7.3.0.N1.0.2.bin

TestFabricpathTopo#test_topo_destroy = 53.35 s = .
TestFabricpathTopo#test_topo_create = 45.40 s = .
TestFabricpathTopo#test_member_vlans_invalid = 49.17 s = .
TestFabricpathTopo#test_member_vlans = 46.26 s = .
TestFabricpathTopo#test_invalid_topo_id = 44.27 s = .

Finished in 238.453413s, 0.0210 runs/s, 0.0671 assertions/s.

5 runs, 16 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /ws/dcherian-sjc/git_projects/develop/cisco-network-node-utils/tests/coverage. 0.0 / 0.0 LOC (100.0%) covered.
# FABRICPATH GLOBAL:

$  minitest ./test_fabricpath_global.rb 10.122.84.136 admin 
Run options: -v -- --seed 29420
# Running:

Node under test:
- name  - n8k-136
- type  - N5K-C56128P
- image - bootflash:///n6000-uk9.7.3.0.N1.0.2.bin

TestFabricpathGlobal#test_mode = 52.92 s = .
TestFabricpathGlobal#test_loadbalance_multicast = 43.98 s = .
TestFabricpathGlobal#test_ttl_unicast = 44.01 s = .
TestFabricpathGlobal#test_global_destroy = 49.41 s = .
TestFabricpathGlobal#test_linkup_delay_all = 45.20 s = .
TestFabricpathGlobal#test_global_create = 44.68 s = .
TestFabricpathGlobal#test_ttl_multicast = 43.88 s = .
TestFabricpathGlobal#test_switch_id = 49.64 s = .
TestFabricpathGlobal#test_graceful_merge = 46.49 s = .
TestFabricpathGlobal#test_aggregate_multicast_routes = 43.74 s = .
TestFabricpathGlobal#test_allocate_delay = 45.68 s = .
TestFabricpathGlobal#test_global_collection_empty = 2.88 s = .
TestFabricpathGlobal#test_loadbalance_unicast = 50.02 s = .
TestFabricpathGlobal#test_transition_delay = 45.66 s = .
TestFabricpathGlobal#test_loadbalance_algorithm = 45.22 s = .

Finished in 653.410273s, 0.0230 runs/s, 0.0398 assertions/s.

15 runs, 26 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /ws/dcherian-sjc/git_projects/develop/cisco-network-node-utils/tests/coverage. 0.0 / 0.0 LOC (100.0%) covered.

dcheriancisco committed:
adding fabricpath topo and fixed feature_set as a class method

dcheriancisco committed:
Fixed minor rubocop errors

mikewiebe commented:
Many of these rubocop failures are due to the fact that the `feature/fabricpath` branch does not have the latest updates from `develop`.  You will need to merge `develop` into `feature/fabricpath` and then sync your fork with `feature/fabricpath` but note that new yaml refactoring changes have also gone into `develop` that will results in some refactoring to this PR.

mikewiebe commented on lib/cisco_node_utils/command_reference_n7k.yaml:
```diff
@@ -114,7 +114,19 @@ fabricpath:
+  ttl_multicast:
+    config_get: "show fabricpath ttl"
+    config_get_token: '/^Multicast TTL:\s+(\d+) hops/'
+    config_set: "%s fabricpath ttl multicast %s"
```
The use of Key-value wildcards is preferred over Printf-style wildcards.

See: https://github.com/cisco/cisco-network-node-utils/blob/develop/docs/README-develop-best-practices.md#yaml8

The following guide goes into more detail on how to use Key-value wild cards, but note this guide is revised in the latest develop branch.

https://github.com/cisco/cisco-network-node-utils/blob/feature/fabricpath/lib/cisco_node_utils/README_YAML.md

chrisvanheuveln commented on tests/test_fabricpath_global.rb:
```diff
@@ -216,4 +216,24 @@ def test_transition_delay
+    @global.ttl_unicast = 40
+    assert_equal(40, @global.ttl_unicast,
+                 'unicast ttl not getting set to 40')
+  end
```
1) Don't hard-code the default value
2) You can remove the duplication with one test method:

```
def test_ttl_multicast_unicast
  @global = FabricpathGlobal.new('default')
  %w(ttl_multicast, ttl_unicast).each do |prop|
    # test default value
    default = @global.send("default_#{prop}")
    @global.send("#{prop}=", default)
    assert_equal(default, @global.send(prop),
                 "#{prop} : Default not set to #{default}")

    # test non-default value
    non_default = default + 8
    @global.send("#{prop}=", non_default)
    assert_equal(non_default, @global.send(prop),
                 "#{prop} : Non-default not set to #{non_default}")
  end
end
```

dcheriancisco replied:
OK I can re-work this.

Thx,
deepak

From: Chris Van Heuveln <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cisco/cisco-network-node-utils <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Wednesday, November 11, 2015 at 7:59 AM
To: cisco/cisco-network-node-utils <cisco-network-node-utils@noreply.github.com<mailto:cisco-network-node-utils@noreply.github.com>>
Cc: Cisco Employee <dcherian@cisco.com<mailto:dcherian@cisco.com>>
Subject: Re: [cisco-network-node-utils] adding fabricpath topo and fixed feature_set as a class method (#75)

In tests/test_fabricpath_global.rbhttps://github.com/cisco/cisco-network-node-utils/pull/75#discussion_r44548286:

> -    assert_equal(32, @global.ttl_multicast,
> -                 'Default multicast ttl not set to 32')
> -    @global.ttl_multicast = 16
> -    assert_equal(16, @global.ttl_multicast,
> -                 'multicast ttl not getting set to 16')
> -  end
>   +
> -  def test_ttl_unicast
> -    @global = FabricpathGlobal.new('default')
> -    # test default value
> -    assert_equal(32, @global.ttl_unicast,
> -                 'Default unicast ttl not set to 32')
> -    @global.ttl_unicast = 40
> -    assert_equal(40, @global.ttl_unicast,
> -                 'unicast ttl not getting set to 40')
> -  end

1) Don't hard-code the default value
2) You can remove the duplication with:

def test_ttl_unicast
  @global = FabricpathGlobal.new('default')
  %w(ttl_multicast, ttl_unicast).each do |prop|
    # test default value
    default = @global.send("default_#{prop}")
    @global.send("#{prop}=", default)
    assert_equal(default, @global.send(prop),
                 "#{prop} : Default not set to #{default}")
    # test non-default value
    non_default = default + 8
    @global.send("#{prop}=", non_default)
    assert_equal(non_default, @global.send(prop),
                 "#{prop} : Non-default not set to #{non_default}")
  end
end

## 

Reply to this email directly or view it on GitHubhttps://github.com/cisco/cisco-network-node-utils/pull/75/files#r44548286.

mikewiebe commented on lib/cisco_node_utils/fabricpath_global.rb:
```diff
@@ -40,15 +40,15 @@ def self.globals
     end
 
-    def fabricpath_feature
+    def self.fabricpath_feature
```
What is the rational behind the change from instance to class method?

dcheriancisco replied:
So that it can be re-used in other classes like interface, vlan, fabricpath_topology etc., which need to check if the fabricpath feature is enabled, instead of repeating the code there. I thought you guys suggested that when I tried to use mixins.

Thx,
deepak

From: Mike Wiebe <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cisco/cisco-network-node-utils <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Wednesday, November 11, 2015 at 8:01 AM
To: cisco/cisco-network-node-utils <cisco-network-node-utils@noreply.github.com<mailto:cisco-network-node-utils@noreply.github.com>>
Cc: Cisco Employee <dcherian@cisco.com<mailto:dcherian@cisco.com>>
Subject: Re: [cisco-network-node-utils] adding fabricpath topo and fixed feature_set as a class method (#75)

In lib/cisco_node_utils/fabricpath_global.rbhttps://github.com/cisco/cisco-network-node-utils/pull/75#discussion_r44548502:

> @@ -40,15 +40,15 @@ def self.globals
>        hash
>      end
> -    def fabricpath_feature
> -    def self.fabricpath_feature

What is the rational behind the change from instance to class method?

## 

Reply to this email directly or view it on GitHubhttps://github.com/cisco/cisco-network-node-utils/pull/75/files#r44548502.

mikewiebe replied:
Ok, sounds good.

dcheriancisco committed:
commit before merge with develop

dcheriancisco committed:
Merge branch 'develop' into feature/fabricpath

dcheriancisco committed:
Merge remote-tracking branch 'upstream/feature/fabricpath' into feature/fabricpath. Add required YAML refactoring code

dcheriancisco committed:
fixed fabricpath topology minitest failure

dcheriancisco commented:
Added re-factored code for new YAML. Fixed Rubocop errors and ran mintests successfully:

# N7K Global and Topo:

$ minitest ./test_fabricpath_global.rb 172.23.145.237 nbv_1234
Run options: -v -- --seed 5739

# Running:

Node under test:
- name  - n7k9
- type  - N7K-C7010
- image - bootflash:///n7000-s2-dk9.7.3.0.OAC.0.8.gbin

TestFabricpathGlobal#test_linkup_delay_all = 23.02 s = .
TestFabricpathGlobal#test_aggregate_multicast_routes = 20.49 s = .
TestFabricpathGlobal#test_allocate_delay = 20.81 s = .
TestFabricpathGlobal#test_global_destroy = 19.44 s = .
TestFabricpathGlobal#test_global_create = 20.26 s = .
TestFabricpathGlobal#test_loadbalance_multicast = 21.13 s = .
TestFabricpathGlobal#test_ttl_multicast = 21.70 s = .
TestFabricpathGlobal#test_loadbalance_unicast = 22.68 s = .
TestFabricpathGlobal#test_global_collection_empty = 1.31 s = .
TestFabricpathGlobal#test_switch_id = 20.61 s = .
TestFabricpathGlobal#test_transition_delay = 20.71 s = .
TestFabricpathGlobal#test_mode = 20.16 s = .
TestFabricpathGlobal#test_loadbalance_algorithm = 22.22 s = .
TestFabricpathGlobal#test_graceful_merge = 22.34 s = .
TestFabricpathGlobal#test_ttl_unicast = 22.63 s = .

Finished in 299.495771s, 0.0501 runs/s, 0.1369 assertions/s.

15 runs, 41 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /ws/dcherian-sjc/git_projects/develop/cisco-network-node-utils/tests/coverage. 0.0 / 0.0 LOC (100.0%) covered.

$ minitest ./test_fabricpath_topology.rb 172.23.145.237 nbv_1234
Run options: -v -- --seed 18440

# Running:

Node under test:
- name  - n7k9
- type  - N7K-C7010
- image - bootflash:///n7000-s2-dk9.7.3.0.OAC.0.8.gbin

TestFabricpathTopo#test_member_vlans_invalid = 20.61 s = .
TestFabricpathTopo#test_invalid_topo_id = 19.14 s = .
TestFabricpathTopo#test_member_vlans = 21.30 s = .
TestFabricpathTopo#test_topo_destroy = 20.63 s = .
TestFabricpathTopo#test_topo_create = 19.58 s = .

Finished in 101.258070s, 0.0494 runs/s, 0.1580 assertions/s.

5 runs, 16 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /ws/dcherian-sjc/git_projects/develop/cisco-network-node-utils/tests/coverage. 0.0 / 0.0 LOC (100.0%) covered.

# N5K global and Topo:

$ minitest ./test_fabricpath_global.rb 10.122.84.136 admin
Run options: -v -- --seed 6792

# Running:

Node under test:
- name  - n8k-136
- type  - N5K-C56128P
- image - bootflash:///n6000-uk9-kickstart.7.3.0.N1.0.2.bin

TestFabricpathGlobal#test_mode = 52.72 s = .
TestFabricpathGlobal#test_aggregate_multicast_routes = 2.94 s = .
TestFabricpathGlobal#test_loadbalance_algorithm = 45.32 s = .
TestFabricpathGlobal#test_ttl_unicast = 2.93 s = .
TestFabricpathGlobal#test_graceful_merge = 49.94 s = .
TestFabricpathGlobal#test_linkup_delay_all = 45.58 s = .
TestFabricpathGlobal#test_ttl_multicast = 3.01 s = .
TestFabricpathGlobal#test_transition_delay = 45.33 s = .
TestFabricpathGlobal#test_global_collection_empty = 2.90 s = .
TestFabricpathGlobal#test_allocate_delay = 45.60 s = .
TestFabricpathGlobal#test_loadbalance_unicast = 51.30 s = .
TestFabricpathGlobal#test_global_destroy = 44.93 s = .
TestFabricpathGlobal#test_switch_id = 45.25 s = .
TestFabricpathGlobal#test_loadbalance_multicast = 2.99 s = .
TestFabricpathGlobal#test_global_create = 44.41 s = .

Finished in 485.161248s, 0.0309 runs/s, 0.0536 assertions/s.

15 runs, 26 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /ws/dcherian-sjc/git_projects/develop/cisco-network-node-utils/tests/coverage. 0.0 / 0.0 LOC (100.0%) covered.

$ minitest ./test_fabricpath_topology.rb 10.122.84.136 admin
Run options: -v -- --seed 26114

# Running:

Node under test:
- name  - n8k-136
- type  - N5K-C56128P
- image - bootflash:///n6000-uk9-kickstart.7.3.0.N1.0.2.bin

TestFabricpathTopo#test_topo_create = 52.19 s = .
TestFabricpathTopo#test_topo_destroy = 46.43 s = .
TestFabricpathTopo#test_member_vlans = 50.89 s = .
TestFabricpathTopo#test_member_vlans_invalid = 44.36 s = .
TestFabricpathTopo#test_invalid_topo_id = 44.19 s = .

Finished in 238.067257s, 0.0210 runs/s, 0.0672 assertions/s.

5 runs, 16 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /ws/dcherian-sjc/git_projects/develop/cisco-network-node-utils/tests/coverage. 0.0 / 0.0 LOC (100.0%) covered.

mikewiebe commented on lib/cisco_node_utils/cmd_ref/fabricpath.yaml:
```diff
@@ -0,0 +1,178 @@
+# fabricpath
+---
+# Fabricpath feature is not available on N3K and N9K
+_exclude: [/N3/, /N9K/]
```
Are you missing the 'K' for N3K?

dcheriancisco replied:
I purposely excluded the K, to allow for more model variations just like N7700 and N7K (there are 2 variations for N7K)

Thx,
deepak

From: Mike Wiebe <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cisco/cisco-network-node-utils <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Friday, November 13, 2015 at 7:02 AM
To: cisco/cisco-network-node-utils <cisco-network-node-utils@noreply.github.com<mailto:cisco-network-node-utils@noreply.github.com>>
Cc: Cisco Employee <dcherian@cisco.com<mailto:dcherian@cisco.com>>
Subject: Re: [cisco-network-node-utils] adding fabricpath topo and fixed feature_set as a class method (#75)

In lib/cisco_node_utils/cmd_ref/fabricpath.yamlhttps://github.com/cisco/cisco-network-node-utils/pull/75#discussion_r44790542:

> @@ -0,0 +1,178 @@
> +# fabricpath
> +---
> +# Fabricpath feature is not available on N3K and N9K
> +_exclude: [/N3/, /N9K/]

Are you missing the 'K' for N3K?

## 

Reply to this email directly or view it on GitHubhttps://github.com/cisco/cisco-network-node-utils/pull/75/files#r44790542.

mikewiebe commented on lib/cisco_node_utils/cmd_ref/fabricpath.yaml:
```diff
@@ -0,0 +1,178 @@
+  kind: boolean
+  config_get: "show run fabricpath"
+  config_get_token: '/^fabricpath multicast aggregate\-routes$/'
+  config_set: "%s fabricpath multicast aggregate-routes"
```
What is the rational for using printf-style vs. key-value wildcard substitution?  Using Key-value wildcards are more flexible in terms of multiple platform support.

https://github.com/cisco/cisco-network-node-utils/blob/develop/lib/cisco_node_utils/cmd_ref/README_YAML.md#wildcard-substitution

dcheriancisco replied:
My code already had a whole bunch of printf-style configs. Wherever there are platform differences, I have tried to use key-value style

From: Mike Wiebe <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cisco/cisco-network-node-utils <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Friday, November 13, 2015 at 7:57 AM
To: cisco/cisco-network-node-utils <cisco-network-node-utils@noreply.github.com<mailto:cisco-network-node-utils@noreply.github.com>>
Cc: Cisco Employee <dcherian@cisco.com<mailto:dcherian@cisco.com>>
Subject: Re: [cisco-network-node-utils] adding fabricpath topo and fixed feature_set as a class method (#75)

In lib/cisco_node_utils/cmd_ref/fabricpath.yamlhttps://github.com/cisco/cisco-network-node-utils/pull/75#discussion_r44797760:

> @@ -0,0 +1,178 @@
> +# fabricpath
> +---
> +# Fabricpath feature is not available on N3K and N9K
> +_exclude: [/N3/, /N9K/]
> +
> +aggregate_multicast_routes:
> -  _exclude:
> -    - /N5K/
> -  kind: boolean
> -  config_get: "show run fabricpath"
> -  config_get_token: '/^fabricpath multicast aggregate-routes$/'
> -  config_set: "%s fabricpath multicast aggregate-routes"

What is the rational for using printf-style vs. key-value wildcard substitution? Using Key-value wildcards are more flexible in terms of multiple platform support.

https://github.com/cisco/cisco-network-node-utils/blob/develop/lib/cisco_node_utils/cmd_ref/README_YAML.md#wildcard-substitution

## 

Reply to this email directly or view it on GitHubhttps://github.com/cisco/cisco-network-node-utils/pull/75/files#r44797760.

mikewiebe replied:
At the very least we need to be consistent, but in general we are migrating all existing code to use the key-value wildcards given that all new providers use that to align with our best practices for future extensibility.  I know it's a bit more work but I think it will pay off in the long run.

mikewiebe commented on lib/cisco_node_utils/cmd_ref/vlan.yaml:
```diff
@@ -11,6 +11,16 @@ create:
+mode:
+  _exclude:
+    - /N9K/
+    - /N3/
```
Missing 'K' here?  Also, you should be able to exclude these using a single line.

`_exclude: [/N3K/, /N9K/]`

dcheriancisco replied:
Since it is a regex, hopefully it will work for all N3 variants. I see that the product id is something like N3064.

From: Mike Wiebe <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cisco/cisco-network-node-utils <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Friday, November 13, 2015 at 8:01 AM
To: cisco/cisco-network-node-utils <cisco-network-node-utils@noreply.github.com<mailto:cisco-network-node-utils@noreply.github.com>>
Cc: Cisco Employee <dcherian@cisco.com<mailto:dcherian@cisco.com>>
Subject: Re: [cisco-network-node-utils] adding fabricpath topo and fixed feature_set as a class method (#75)

In lib/cisco_node_utils/cmd_ref/vlan.yamlhttps://github.com/cisco/cisco-network-node-utils/pull/75#discussion_r44798265:

> @@ -11,6 +11,16 @@ create:
>  destroy:
>    config_set: "no vlan %s"
> 
> +mode:
> -  _exclude:
> -    - /N9K/
> -    - /N3/

Missing 'K' here? Also, you should be able to exclude these using a single line.

_exclude: [/N3K/, /N9K/]

## 

Reply to this email directly or view it on GitHubhttps://github.com/cisco/cisco-network-node-utils/pull/75/files#r44798265.

mikewiebe replied:
Ok

mikewiebe commented on lib/cisco_node_utils/fabricpath_global.rb:
```diff
@@ -35,20 +35,20 @@ def initialize(name, instantiate=true)
       hash = {}
       is_fabricpath_feature = config_get('fabricpath', 'feature')
-      return hash if (:enabled != is_fabricpath_feature.first.to_sym)
+      return hash if (:enabled != is_fabricpath_feature.to_sym)
```
Just a thought.  Could you rewrite the feature yaml entry to add `default_value` and `kind` features and check for enabled explicitly?

```
feature:
  default_value: false
  kind: boolean:
  config_get: "show feature-set"
  config_get_token: '/^fabricpath[\s]+[\d]+[\s]+(enabled)/'
  config_set: "%s feature-set fabricpath"
```

Then you could collapse the feature enabled check into a single line like this...

`return hash unless config_get('fabricpath', 'feature')`

dcheriancisco replied:
Update11/14:
Actually, I can't do this. Since the config_get() is used to check if the feature is installed or un-installed as well in a later class method. It is not a simple boolean.

Yes, I could do that. This was leftover code when I had mixin. The feature check method was not working well with the mixin.

From: Mike Wiebe <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cisco/cisco-network-node-utils <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Friday, November 13, 2015 at 8:26 AM
To: cisco/cisco-network-node-utils <cisco-network-node-utils@noreply.github.com<mailto:cisco-network-node-utils@noreply.github.com>>
Cc: Cisco Employee <dcherian@cisco.com<mailto:dcherian@cisco.com>>
Subject: Re: [cisco-network-node-utils] adding fabricpath topo and fixed feature_set as a class method (#75)

In lib/cisco_node_utils/fabricpath_global.rbhttps://github.com/cisco/cisco-network-node-utils/pull/75#discussion_r44801697:

> @@ -35,20 +35,20 @@ def initialize(name, instantiate=true)
>      def self.globals
>        hash = {}
>        is_fabricpath_feature = config_get('fabricpath', 'feature')
> -      return hash if (:enabled != is_fabricpath_feature.first.to_sym)
> -      return hash if (:enabled != is_fabricpath_feature.to_sym)

Just a thought. Could you rewrite the feature yaml entry to add default_value and kind features and check for enabled explicitly?

feature:
  default_value: false
  kind: boolean:
  config_get: "show feature-set"
  config_get_token: '/^fabricpath[\s]+[\d]+[\s]+(enabled)/'
  config_set: "%s feature-set fabricpath"

Then you could collapse the feature enabled check into a single line like this...

return hash unless config_get('fabricpath', 'feature')

## 

Reply to this email directly or view it on GitHubhttps://github.com/cisco/cisco-network-node-utils/pull/75/files#r44801697.

mikewiebe commented on lib/cisco_node_utils/fabricpath_global.rb:
```diff
@@ -103,13 +104,11 @@ def my_munge(property, set_val)
 
     def aggregate_multicast_routes=(val)
-      if (val == true)
+      if val == true
```
Suggest the following:

```
val ? state = '' : state = 'no'
config_set('fabricpath', 'aggregate_multicast_routes', state)
```

mikewiebe commented on lib/cisco_node_utils/fabricpath_global.rb:
```diff
@@ -142,15 +139,15 @@ def default_allocate_delay
+      graceful_merge_conf = config_get('fabricpath', 'graceful_merge')
+      # opposite meaning with the cli
+      return true if graceful_merge_conf.nil?
+      graceful_merge_conf == true ? false : true
```
Suggest: `graceful_merge_conf ? false : true` instead.

dcheriancisco replied:
Ok agree

From: Mike Wiebe <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cisco/cisco-network-node-utils <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Friday, November 13, 2015 at 8:38 AM
To: cisco/cisco-network-node-utils <cisco-network-node-utils@noreply.github.com<mailto:cisco-network-node-utils@noreply.github.com>>
Cc: Cisco Employee <dcherian@cisco.com<mailto:dcherian@cisco.com>>
Subject: Re: [cisco-network-node-utils] adding fabricpath topo and fixed feature_set as a class method (#75)

In lib/cisco_node_utils/fabricpath_global.rbhttps://github.com/cisco/cisco-network-node-utils/pull/75#discussion_r44803307:

> ```
>  end
> ```
> -    def auto_switch_id
> -      val = config_get('fabricpath', 'auto_switch_id')
> -      val.first.to_i
> -    def graceful_merge
> -      graceful_merge_conf = config_get('fabricpath', 'graceful_merge')
> -      # opposite meaning with the cli
> -      return true if graceful_merge_conf.nil?
> -      graceful_merge_conf == true ? false : true

Suggest: graceful_merge_conf ? false : true instead.

## 

Reply to this email directly or view it on GitHubhttps://github.com/cisco/cisco-network-node-utils/pull/75/files#r44803307.

mikewiebe commented on lib/cisco_node_utils/fabricpath_global.rb:
```diff
@@ -209,9 +202,9 @@ def default_linkup_delay_always
-      enabled = config_get('fabricpath', 'linkup_delay_enable')
-      return default_linkup_delay_enable if enabled.nil?
-      enabled.first.strip == 'Enabled' ? true : false
+      delay = config_get('fabricpath', 'linkup_delay_enable')
```
Why does this property need to use`auto_default`?

dcheriancisco replied:
This is a gotcha with boolean kind and auto-default. The missing match causes a nil and therefore forces the use of default value. If the default value is "true" we get an undesired result.

This is the complete code (as seen below, we want to return false for no match. NOTE: the command used is not a "show run"):
    def linkup_delay_enable
      delay = config_get('fabricpath', 'linkup_delay_enable')
      return false if delay.nil?
      delay
    end

From: Mike Wiebe <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cisco/cisco-network-node-utils <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Friday, November 13, 2015 at 8:41 AM
To: cisco/cisco-network-node-utils <cisco-network-node-utils@noreply.github.com<mailto:cisco-network-node-utils@noreply.github.com>>
Cc: Cisco Employee <dcherian@cisco.com<mailto:dcherian@cisco.com>>
Subject: Re: [cisco-network-node-utils] adding fabricpath topo and fixed feature_set as a class method (#75)

In lib/cisco_node_utils/fabricpath_global.rbhttps://github.com/cisco/cisco-network-node-utils/pull/75#discussion_r44803678:

> @@ -209,9 +202,9 @@ def default_linkup_delay_always
>      end
> 
> ```
>  def linkup_delay_enable
> ```
> -      enabled = config_get('fabricpath', 'linkup_delay_enable')
> -      return default_linkup_delay_enable if enabled.nil?
> -      enabled.first.strip == 'Enabled' ? true : false
> -      delay = config_get('fabricpath', 'linkup_delay_enable')

Why does this property need to useauto_default?

## 

Reply to this email directly or view it on GitHubhttps://github.com/cisco/cisco-network-node-utils/pull/75/files#r44803678.

mikewiebe commented on lib/cisco_node_utils/fabricpath_global.rb:
```diff
@@ -231,8 +224,7 @@ def default_linkup_delay_enable
       algo = config_get('fabricpath', 'loadbalance_algorithm')
-      return default_loadbalance_algorithm if algo.nil?
-      algo.first.strip.downcase
+      algo.downcase
```
Should be able to just do `config_get('fabricpath', 'loadbalance_algorithm').downcase`

dcheriancisco replied:
ok

From: Mike Wiebe <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cisco/cisco-network-node-utils <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Friday, November 13, 2015 at 8:42 AM
To: cisco/cisco-network-node-utils <cisco-network-node-utils@noreply.github.com<mailto:cisco-network-node-utils@noreply.github.com>>
Cc: Cisco Employee <dcherian@cisco.com<mailto:dcherian@cisco.com>>
Subject: Re: [cisco-network-node-utils] adding fabricpath topo and fixed feature_set as a class method (#75)

In lib/cisco_node_utils/fabricpath_global.rbhttps://github.com/cisco/cisco-network-node-utils/pull/75#discussion_r44803775:

> @@ -231,8 +224,7 @@ def default_linkup_delay_enable
> 
> ```
>  def loadbalance_algorithm
>    algo = config_get('fabricpath', 'loadbalance_algorithm')
> ```
> -      return default_loadbalance_algorithm if algo.nil?
> -      algo.first.strip.downcase
> -      algo.downcase

Should be able to just do config_get('fabricpath', 'loadbalance_algorithm').downcase

## 

Reply to this email directly or view it on GitHubhttps://github.com/cisco/cisco-network-node-utils/pull/75/files#r44803775.

mikewiebe commented on lib/cisco_node_utils/fabricpath_global.rb:
```diff
@@ -421,11 +389,47 @@ def transition_delay=(val)
+    end
+
+    def ttl_multicast=(val)
+      if val == ''
```
```
val=='' ? state = '' : state = 'no'
config_set('fabricpath', 'ttl_multicast', state, val)
```

Same for `ttl_unicast=`

dcheriancisco committed:
Addressed review comments

mikewiebe commented:
:+1:
